---
title: "test_alns_with_gaps.Rmd"
author: "Janet Young\n"
date: "`r Sys.Date()`\n"
output: github_document
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(tidyverse)
library(kableExtra)
# load MKfunctions.R Also loads the Biostrings package. 
source(here("scripts/MKfunctions.R"))
```



I was getting an error when all members of one population have a gap for a whole codon. This showed up when Sage used my code to analyze the mod(mdg4) gene.

For testing, I made a small example alignment, that has a gap in every pop1 sequences at some positions (file = `test_data/testData_for_troubleshooting/test_otherMiscProblemAlns/test_onePopAllGaps.fa`)


```{r}
## example alignment WITHOUT gaps in all members of one population
alnFile <- here("test_data/MKTwebsiteExample/MKTwebsite_testAln.fa")
aln <- readBStringSet(alnFile)
names(aln) <- sapply(strsplit(names(aln), " "), "[[", 1)
alnSpecies <- gsub("\\d","",names(aln))
alnPopulations <- split(names(aln), alnSpecies)

## example alignment WITH gaps in all members of one population
alnWithGapsFile <- here("test_data/testData_for_troubleshooting/test_otherMiscProblemAlns/test_onePopAllGaps.fa")
alnWithGaps <- readBStringSet(alnWithGapsFile)
names(alnWithGaps) <- sapply(strsplit(names(alnWithGaps), " "), "[[", 1)
alnWithGapsSpecies <- gsub("\\d","",names(alnWithGaps))
alnWithGapsPopulations <- split(names(alnWithGaps), alnWithGapsSpecies)
```


I tracked down the error, to a call within `doMKtest()`, to the `getCodonChangeCounts()` function (I've fixed it now). This code was the minimal example to show the error:

```{r}
getCodonChangeCounts("---","CTA")
```

Run `doMKtest` on alignment without gaps:

```{r}
MKresults_noGaps <- doMKtest(alnFile, 
                      pop1seqs=alnWithGapsPopulations[["pongo"]],
                      pop2seqs=alnWithGapsPopulations[["trachy"]],
                      quiet=TRUE)
```
Run `doMKtest` on alignment with gaps:

```{r}
MKresults <- doMKtest(alnWithGapsFile, 
                      pop1seqs=alnWithGapsPopulations[["pongo"]],
                      pop2seqs=alnWithGapsPopulations[["trachy"]],
                      quiet=TRUE)
```
Other than the `input` column (tells us the input alignment file name), the results are identical. That's good, because the codons where I introduced artificial gaps did not contain any evolutionary changes.

```{r}
identical(MKresults_noGaps[["summary"]] %>% dplyr::select(-input), 
          MKresults[["summary"]] %>% dplyr::select(-input))
```


Show MK results (alignment with gaps):

```{r}
MKresults[["summary"]] %>% 
    dplyr::select(-input) %>% 
    kable() %>% 
    kable_styling()
```

Show positions table for one of the gap codons:

```{r}
MKresults[["positions"]] %>% 
    # mutate is because kable turns hyphens into bullets
    # https://github.com/haozhu233/kableExtra/issues/223
    mutate(pop1_anc=gsub("-","--",pop1_anc)) %>% 
    mutate(pop2_anc=gsub("-","--",pop2_anc)) %>% 
    as_tibble() %>% 
    filter(pos %in% 16:18) %>% 
    kable() %>% 
    kable_styling()
```




# Finished

```{r}
sessionInfo()
```



